### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by sonnysideup**
- adding basic types
- implementing basic vagrant commands (starting off simple here)

### CODE REVIEW

**Comment by sonnysideup** [comment]
This still needs a basic README but the logic is largely 99% here for what we need.

File: machine.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	case "stuck":
+		return Stuck
+	case "inaccessible":
+		return Inaccessible
+	case "gurumeditation":
+		return GuruMeditation
+	default:
+		return Unknown
+	}
+}
```
Comment (steved) [inline machine.go]: Could probably clean up `ToMachineState` and `String` a bit with some sort of constant map like you use in the test cases, but it's pretty minor.

Comment (sonnysideup) [reply machine.go]: Interestingly, we would either need to use two maps, or a map of maps/structs. Did you noticed that string-to-state and state-to-string conversion is not symmetric?

```go
// used to provide better context in an error/log message
NotCreated.String() // => "NotCreated"

// converts string provided by Vagrant into a state constant
ToMachineState("not_created") // => NotCreated
```

Were you thinking of two maps like the following?
```go
var (
	msStrings = []string{
		"Unknown", "Aborted", "GuruMeditation", "Inaccessible", "NotCreated",
		"Paused", "PowerOff", "Stopping", "Running", "Saving", "Saved", "Stuck"}
	msMap = map[string]MachineState{
		"running": Running,
		"not_created": NotCreated,
		....
	}
)
```

File: machine.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	mType     string
+	data      []string
+}
+
+// parseMachineReadable converts machine-readable output into a slice of machineOutputEntry.
+func parseMachineReadable(machineOut []byte) (entries []machineOutputEntry, err error) {
+	scanner := bufio.NewScanner(strings.NewReader(string(machineOut)))
+	for scanner.Scan() {
+		line := scanner.Text()
+		row := strings.Split(line, ",")
```
[the remark this answers is not in the document]
Comment (sonnysideup) [reply machine.go]: Great question. The format is such that `%!(VAGRANT_COMMA)` will only appear in the `data` field; see https://www.vagrantup.com/docs/cli/machine-readable.html#format.

I should probably add a comment to the code with a link to this documentation.

File: wrapper.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+// New creates a new Vagrant CLI wrapper.
+func New() Vagrant {
+	logger := log.New()
+
+	return wrapper{
+		executable: binary,
+		logger:     logger,
+		runner:     command.ShellRunner{},
```
[the remark this answers is not in the document]
Comment (sonnysideup) [reply wrapper.go]: Would you please say more here. Are you suggesting an `init` function in this file or in `command/runner.go`?

File: wrapper.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	return err
+}
+
+// Destroy stops the running guest machines and destroys all of the resources created during the creation process.
+func (w wrapper) Destroy() error {
+	out, err := w.exec("destroy", "--force")
+	if err == nil {
+		w.info(out)
+	}
+	return err
```
[the remark this answers is not in the document]
Comment (sonnysideup) [reply wrapper.go]: This makes sense given that `up`, `halt` and `destroy` all use the same code. I could see `suspend`, `validate` and other commands using the exact same code if we ever need to add them.

File: wrapper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (steved) [reply wrapper.go]: In `command/runner.go`. Something like `NewShellRunner` to abstract the struct from here.

File: machine.go (hunk 1 above)
Comment (steved) [reply machine.go]: I think something like this:
```golang
package main

import (
	"fmt"
)

type state int

const (
	// Unknown represents any unhandled state.
	Unknown state = iota
)

type MachineState struct {
	state
	name string
}

var (
	machineStates = map[string]MachineState{"unknown": {Unknown, "Unknown"}}
)

func (m MachineState) String() string {
	return m.name
}

func ToMachineState(s string) MachineState {
	return machineStates[s]
}

func main() {
	unknown := ToMachineState("unknown")
	fmt.Printf("%d, %s", unknown.state, unknown)
}
```

File: wrapper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sonnysideup) [reply wrapper.go]: I guess my thought here is that we don't really do anything special to instantiate a `ShellRunner` so a func to create one currently feels superfluous. If the setup ever becomes more involved, then I'd probably agree here.

File: machine.go (hunk 1 above)
Comment (sonnysideup) [reply machine.go]: I think nesting a `state` inside of a `MachineState` that is nested in a `MachineStatus` could produce an odd API:

```go
statuses, _ := wrapper.Status()
for _, status := statuses {
  switch status.State.state {
  case SomeState:
    fmt.Println("State.state seems clunky")
  }
}
```

If I just use a different data structure, then this is a non-issue.

**Pull request merged**